yjit/bindgen/src/main.rs

```diff
@@ -61,72 +72,95 @@ fn main() {
         .allowlist_var("rb_cFloat")
         .allowlist_var("rb_cString")
 
+        // From vm_callinfo.h
         .allowlist_type("VM_CALL.*") // This doesn't work, possibly due to the odd structure of the #defines
         .allowlist_type("vm_call_flag_bits") // So instead we include the other enum and do the bit-shift ourselves
 
+        // From vm_core.h
+        .allowlist_var("VM_BLOCK_HANDLER_NONE")
+        .allowlist_var("VM_ENV_FLAG_.*")
+
+        // From include/ruby/internal/intern/range.h
         .allowlist_function("rb_range_new")
 
+        // From internal/string.h
         .allowlist_function("rb_ec_str_resurrect")
         .allowlist_function("rb_str_concat_literals")
 
         // `ruby_value_type` is a C enum and this stops it from
         // prefixing all the members with the name of the type
         .prepend_enum_name(false)
         .translate_enum_integer_types(true) // so we get fixed width Rust types for members
+        // From include/ruby/internal/value_type.h
         .allowlist_type("ruby_value_type") // really old C extension API
 
+        // Autogenerated into id.h
         .allowlist_type("ruby_method_ids")
 
-        .allowlist_function("rb_callable_method_entry")
+        // From method.h
         .allowlist_type("rb_method_visibility_t")
         .allowlist_type("rb_method_type_t")
         .allowlist_type("method_optimized_type")
 
-        // Constants defined in vm_core.h
+        // Opaque types from method.h
+        .blocklist_type("rb_method_cfunc_t")
+
+        // From vm_core.h
         .allowlist_type("ruby_basic_operators")
         .allowlist_var(".*_REDEFINED_OP_FLAG")
         .allowlist_type("rb_num_t")
+        .allowlist_function("rb_callable_method_entry")
 
-        .allowlist_function("rb_iseq_(get|set)_yjit_payload")
+        // Opaque types from vm_core.h
+        .allowlist_type("IVC") // pointer to iseq_inline_iv_cache_entry
+        .opaque_type("iseq_inline_iv_cache_entry") // inline cache entry
+        .blocklist_type("iseq_inline_iv_cache_entry")
+        .blocklist_type("rb_call_data")
+        .opaque_type("rb_call_data")
+        .blocklist_type("rb_execution_context_.*")
+        .opaque_type("rb_execution_context_.*")
+        .blocklist_type("rb_callable_method_entry_t")
+        .opaque_type("rb_callable_method_entry_t")
 
+        // From yjit.c
+        .allowlist_function("rb_iseq_(get|set)_yjit_payload")
         .allowlist_function("rb_iseq_pc_at_idx")
         .allowlist_function("rb_iseq_opcode_at_pc")
+        .allowlist_function("rb_yjit_mark_writable")
+        .allowlist_function("rb_yjit_mark_executable")
+        .allowlist_function("rb_yjit_get_page_size")
```

Comment:
I think we could allowlist `rb_yjit_*` because those are all our functions.

Reply:
Doing that the obvious way causes bindgen to emit a ton of bitfield-access code in Rust (I've included a small excerpt below.) I don't think we can manually blocklist the __Bindgen structs since we're already doing that with all types starting with "__" and it's clearly not blocking them.

I'll see if I can find a way to get it not to do that. I'm worried that we'll have to blocklist every bitfield structure it sees referenced -- I'm pretty sure that's why we blocklist some of the opaque CRuby types.

~~~
#[repr(C)]
#[derive(Copy, Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
impl<Storage> __BindgenBitfieldUnit<Storage> {
    #[inline]
    pub const fn new(storage: Storage) -> Self {
        Self { storage }
    }
}
impl<Storage> __BindgenBitfieldUnit<Storage>
where
    Storage: AsRef<[u8]> + AsMut<[u8]>,
{
    #[inline]
    pub fn get_bit(&self, index: usize) -> bool {
        debug_assert!(index / 8 < self.storage.as_ref().len());
        let byte_index = index / 8;
        // ... lots more code follows ...
~~~